catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java

```diff
@@ -98,9 +122,77 @@ public void init() {
         "/cql",
         APPLICATION_JSON,
         (req, res) -> {
-          CqlRequest cqlRequest = mapper.readValue(util.safeGetBody(req), CqlRequest.class);
+          String transformerId = req.queryParams("transform");
+          String body = util.safeGetBody(req);
+          CqlRequest cqlRequest = mapper.readValue(body, CqlRequest.class);
           CqlQueryResponse cqlQueryResponse = executeCqlQuery(cqlRequest);
-          return mapper.toJson(cqlQueryResponse);
+
+          if (StringUtils.isEmpty(transformerId)) {
+            LOGGER.trace("Returning cql query response.");
+            return mapper.toJson(cqlQueryResponse);
+          }
+          for (ServiceReference<QueryResponseTransformer> queryResponseTransformer :
+              queryResponseTransformers) {
+
+            LOGGER.trace("Finding transformer to transform query response.");
+
+            String id = (String) queryResponseTransformer.getProperty("id");
+
+            if (!id.equals(transformerId)) {
+              continue;
+            }
+
+            BinaryContent content =
+                bundleContext
+                    .getService(queryResponseTransformer)
+                    .transform(this.queryResponse, new HashMap<>());
+
+            String mimeType = (String) queryResponseTransformer.getProperty("mime-type");
+            MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+            String fileExt = allTypes.forName(mimeType).getExtension();
+
+            res.status(200);
+
+            String acceptEncoding = req.headers("Accept-Encoding");
+
+            boolean shouldGzip =
+                StringUtils.isNotBlank(acceptEncoding)
+                    && acceptEncoding.toLowerCase().contains("gzip");
+
+            if (!shouldGzip) {
+              LOGGER.trace("Request header does not contain gzip.");
+            }
+
+            res.type(mimeType);
+            String attachment =
+                String.format("attachment;filename=export-%s%s", Instant.now().toString(), fileExt);
+            res.header("Content-Disposition", attachment);
+
+            if (shouldGzip) {
+              res.raw().addHeader("Content-Encoding", "gzip");
+            }
+
+            try (OutputStream servletOutputStream = res.raw().getOutputStream();
+                InputStream resultStream = content.getInputStream()) {
+              if (shouldGzip) {
+                try (OutputStream gzipServletOutputStream =
+                    new GZIPOutputStream(servletOutputStream)) {
+                  IOUtils.copy(resultStream, gzipServletOutputStream);
+                }
+              } else {
+                IOUtils.copy(resultStream, servletOutputStream);
+              }
+            }
+
+            LOGGER.trace(String.format("Response sent in %s file format.", fileExt));
+
+            return "";
```

Comment:
❓ Why return the empty string?

Reply:
the return value is what is included in the body, in the case an attachment is made the body is empty

Reply:
as @rzwiefel mentioned it is odd to have the body include return data in one case but as an attachment in another. but that is also what the team recommended, I would like to hear some input on that. I don't believe the results can be returned in the body instead of an attachment since the attachment is the data that the UI download as an external file.